src/i18_bluesky/plans/curve_fitting.py

```diff
@@ -0,0 +1,160 @@
+from inspect import signature
+
+from scipy.optimize import curve_fit
+from bluesky.callbacks.core import CollectThenCompute
+import numpy as np
+import matplotlib.pyplot as plt
+
+class FitCurves(CollectThenCompute):
+    """
+        Callback listener that processes collected documents and
+        fits detector data with curve :
+        <li>Single curve for 1-dimensional line scan,
+        <li> N curves for grid scans with shape NxM (M points per curve).
+
+        Uses scipy curve_fit function for curve fitting
+        fit_function -> function to be used during fitting
+        fit_bounds -> range for each parameter to be used when fitting.
+         A tuple of (min, max) value for each parameter.
+         e.g. for parameters a, b,c : ( (min a, max a), (min b, max b), (min c, max c))
+    """
+
+    def __init__(self):
+        super(FitCurves, self).__init__()
+        self.fit_function = None
+        self.fit_bounds = None
+        self.results = []
+        self.transform_function = None
+        self.bounds = None # static bounds
+        self.bounds_provider = None # function that provides bounds based on x and y values
+
+    def start(self, doc):
+        self.results = []
+        self.reset()
+        super().start(doc)
+
+    def do_fitting(self, xvals, yvals):
+        bounds = None
+        if self.bounds is not None:
+            bounds = self.bounds
+        if self.bounds_provider is not None:
+            bounds = self.bounds_provider(xvals, yvals)
+            print("Bounds from {} : {}".format(self.bounds_provider.__name__, bounds))
+
+        if bounds is None:
+            return curve_fit(self.fit_function, xvals, yvals)
+        else:
+            return curve_fit(self.fit_function, xvals, yvals, bounds=bounds)
+
+    def determine_scan_shape(self):
+        # Extract information about scan shape from start document :
+        scan_shape = self._start_doc.get('shape')
+        if scan_shape is None:
+            scan_shape = [self._start_doc['num_points']]
+        return scan_shape
+
+    def extract_data(self):
+        """ Extract the x and y values (i.e. position of motor being moved and detector readout)
+        from the event documents"""
```

Comment:
should: define the output format - is it a tuple of lists, dictionaries, ndarrays?